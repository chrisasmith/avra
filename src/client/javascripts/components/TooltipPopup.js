import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Popover, Overlay } from 'react-bootstrap'

const propTypes = {
  target: PropTypes.object,
  show: PropTypes.bool,
  children: PropTypes.element,
  onClickOutside: PropTypes.func.isRequired,
  direction: PropTypes.string,
}

class TooltipPopup extends Component {
  constructor(props, context) {
    super(props, context)

    this.popup = React.createRef()
    this.onGlobalClick = this.onGlobalClick.bind(this)

    this.state = {
      show: true,
    }
  }

  componentWillReceiveProps(next) {
    if (next.show && !this.props.show) {
      // add the global click listener
      document.addEventListener('click', this.onGlobalClick, false)
    }

    if (!next.show && this.props.show) {
      document.removeEventListener('click', this.onGlobalClick, false)
    }
  }

  onGlobalClick(e) {
    // bail if this is a detached element
    // this can happen if we removed the element we clicked on, generally
    // that means we didn't click outside the popup
    if (!document.body.contains(e.target)) {
      return
    }

    // check to see if we contain this click
    const domElement = this.popup.current
    if (domElement && !domElement.contains(e.target) && this.props.show) {
      this.props.onClickOutside()
    }
  }

  render() {
    const { target, show, children, direction } = this.props

    return (
      <div ref={this.popup}>
        <Overlay
          container={this}
          target={target}
          placement={direction || 'bottom'}
          show={show}
        >
          <Popover id={`popover-positioned-${direction || 'bottom'}`}>
            {children}
          </Popover>
        </Overlay>
      </div>
    )
  }
}

TooltipPopup.propTypes = propTypes
export default TooltipPopup
