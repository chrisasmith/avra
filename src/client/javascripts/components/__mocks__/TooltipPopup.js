import React from 'react'

// eslint-disable-next-line react/prop-types
const TooltipPopup = ({ children }) => (<div>{children}</div>)

export default TooltipPopup
