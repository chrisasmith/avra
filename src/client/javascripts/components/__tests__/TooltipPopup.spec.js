import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import TooltipPopup from '../TooltipPopup'

describe('TooltipPopup Component', () => {
  it('Should show', () => {
    const wrapper = shallow(<TooltipPopup
      target={{}}
      onClickOutside={jest.fn()}
    >
      <div>hi</div>
    </TooltipPopup>)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show', () => {
    const wrapper = shallow(<TooltipPopup
      target={{}}
      onClickOutside={jest.fn()}
      show
    >
      <div>hi</div>
    </TooltipPopup>)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
