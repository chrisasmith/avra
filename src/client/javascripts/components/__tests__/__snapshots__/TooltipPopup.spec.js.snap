// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipPopup Component Should show 1`] = `
<div>
  <Overlay
    container={
      TooltipPopup {
        "context": Object {},
        "onGlobalClick": [Function],
        "popup": Object {
          "current": null,
        },
        "props": Object {
          "children": <div>
            hi
          </div>,
          "onClickOutside": [MockFunction],
          "target": Object {},
        },
        "refs": Object {},
        "setState": [Function],
        "state": Object {
          "show": true,
        },
        "updater": Updater {
          "_callbacks": Array [],
          "_renderer": ReactShallowRenderer {
            "_context": Object {},
            "_element": <TooltipPopup
              onClickOutside={[MockFunction]}
              target={Object {}}
            >
              <div>
                hi
              </div>
            </TooltipPopup>,
            "_forcedUpdate": false,
            "_instance": [Circular],
            "_newState": null,
            "_rendered": <div>
              <Overlay
                container={[Circular]}
                placement="bottom"
                rootClose={false}
                show={false}
                target={Object {}}
                transition={[Function]}
              >
                <ForwardRef(Bootstrap(Popover))
                  id="popover-positioned-bottom"
                >
                  <div>
                    hi
                  </div>
                </ForwardRef(Bootstrap(Popover))>
              </Overlay>
            </div>,
            "_rendering": false,
            "_updater": [Circular],
          },
        },
        Symbol(enzyme.__setState__): [Function],
      }
    }
    placement="bottom"
    rootClose={false}
    show={false}
    target={Object {}}
    transition={[Function]}
  >
    <ForwardRef(Bootstrap(Popover))
      id="popover-positioned-bottom"
    >
      <div>
        hi
      </div>
    </ForwardRef(Bootstrap(Popover))>
  </Overlay>
</div>
`;

exports[`TooltipPopup Component Should show 2`] = `
<div>
  <Overlay
    container={
      TooltipPopup {
        "context": Object {},
        "onGlobalClick": [Function],
        "popup": Object {
          "current": null,
        },
        "props": Object {
          "children": <div>
            hi
          </div>,
          "onClickOutside": [MockFunction],
          "show": true,
          "target": Object {},
        },
        "refs": Object {},
        "setState": [Function],
        "state": Object {
          "show": true,
        },
        "updater": Updater {
          "_callbacks": Array [],
          "_renderer": ReactShallowRenderer {
            "_context": Object {},
            "_element": <TooltipPopup
              onClickOutside={[MockFunction]}
              show={true}
              target={Object {}}
            >
              <div>
                hi
              </div>
            </TooltipPopup>,
            "_forcedUpdate": false,
            "_instance": [Circular],
            "_newState": null,
            "_rendered": <div>
              <Overlay
                container={[Circular]}
                placement="bottom"
                rootClose={false}
                show={true}
                target={Object {}}
                transition={[Function]}
              >
                <ForwardRef(Bootstrap(Popover))
                  id="popover-positioned-bottom"
                >
                  <div>
                    hi
                  </div>
                </ForwardRef(Bootstrap(Popover))>
              </Overlay>
            </div>,
            "_rendering": false,
            "_updater": [Circular],
          },
        },
        Symbol(enzyme.__setState__): [Function],
      }
    }
    placement="bottom"
    rootClose={false}
    show={true}
    target={Object {}}
    transition={[Function]}
  >
    <ForwardRef(Bootstrap(Popover))
      id="popover-positioned-bottom"
    >
      <div>
        hi
      </div>
    </ForwardRef(Bootstrap(Popover))>
  </Overlay>
</div>
`;
